### PR TITLE
feat(website): Table of Contents Component

### PR DESCRIPTION
**Pull request by richbachman**
- [x] Added table of contents component excluding `h1` elements
- [x] Added react-scrollspy and types
- [x] Added TOC to doc pages
- [x] Added `contentwrapper` to `PasteMDXProvider`

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Comment by SiTaggart** [comment]
Is there any way to get this to magically appear on every MDX page via a template, rather than having to manually add it yourself?

**Comment by richbachman** [comment]
I'll dig into that. Adding to all page is kind of a pain, and I clearly missed adding this everywhere.

**Comment by aayushpi** [comment]
That seems like a separate ticket

**Comment by richbachman** [comment]
@aayushpi Agreed. Its going to be a heavier lift to alter the templates.

**Comment by richbachman** [comment]
I also added a TODO to figure out how to handle the changelogs on the various pages. Since those are imported, the headings don't show up in the `mdx` header query. So we need to work out a way to only show the changelog TOC menu item on pages that have a changelog. I've added a ticket to figure that out.

**Comment by SiTaggart** [comment]
If it’s not too much work let’s do it. I don’t want us creating a bunch of work for ourselves in our future or for day to day maintenance

**Comment by richbachman** [comment]
@SiTaggart I think its definitely something we need to do. I think it may be more work than this ticket covers. We'll need to:
- Restructure the `mdx` template
- Rework the `pageQuery` so its dynamic by slug or something like that and make sure it works with the `ComponentHeader` and `TableOfContents`
- Rework to Do/Dont component images so they pull from somewhere other than the `pageQuery`
- Not break anything

I propose creating a new ticket for that, and prioritize so we're only reconfiguring the pages we have now.

Thoughts?

File: packages/paste-website/src/components/shortcodes/table-of-contents/index.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import {TableOfContentsProps} from './types';
+import {StyledWrapper, StyledScrollSpy, StyledListItem} from './styles';
+import {slugify} from '../../../utils/RouteUtils';
+
+const TableOfContents: React.FC<TableOfContentsProps> = data => {
+  if (data == null || data.data == null) {
+    return null;
+  }
+
+  const headings = data.data;
```
Comment (TheSisb) [inline packages/paste-website/src/components/shortcodes/table-of-contents/index.tsx]: `data.data` is kind of weird.  Does the first `data` have a more accurate name by any chance?

File: packages/paste-website/src/pages/tokens/design-tokens-package.mdx [hunk 2]
```diff
@@ -91,8 +101,10 @@ As you use Paste, you'll likely encounter things that _don't seem right_. Please
 
 ***
 
-<Box marginTop="space90">
+<Box marginTop="space90" id="changelog">
```
Comment (TheSisb) [inline packages/paste-website/src/pages/tokens/design-tokens-package.mdx]: Why does this need an id?

File: packages/paste-website/src/components/shortcodes/table-of-contents/index.tsx (hunk 1 above)
Comment (SiTaggart) [reply packages/paste-website/src/components/shortcodes/table-of-contents/index.tsx]: This is props, right?

File: packages/paste-website/src/components/shortcodes/table-of-contents/styles.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  margin-top: ${marginTopPartial};
+  margin-left: ${marginLeftPartial};
+  padding-left: ${themeGet('space.space30')};
+  font-size: ${fontSizePartial};
+  color: ${themeGet('colors.colorGray80')};
+  border-left-width: 2px;
+  border-left-style: solid;
+  border-left-color: transparent;
+
+  a {
```
Comment (SiTaggart) [inline packages/paste-website/src/components/shortcodes/table-of-contents/styles.ts]: Let's avoid descendants like this. You are trying to override the styles of the Anchor component and not honouring component encapsulation.

I'd actually like to request a design change @two24studios, our links / clickable thing are blue, can we keep them blue so it's consistent? Then we don't even need to override the styles

File: packages/paste-website/src/components/shortcodes/table-of-contents/styles.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    color: ${themeGet('colors.colorGray80')};
+  }
+
+  &.is-current {
```
Comment (SiTaggart) [inline packages/paste-website/src/components/shortcodes/table-of-contents/styles.ts]: This is kind of grim, and it's a fault of the library. Generally, components modifying their children is not a good idea. 

We don't have a choice here, but I'm just voicing my ick 😉

**Comment by SiTaggart** [comment]
> I propose creating a new ticket for that, and prioritize so we're only reconfiguring the pages we have now.

OK, I buy it

File: packages/paste-website/src/components/shortcodes/table-of-contents/styles.ts (hunk 4 above)
Comment (TheSisb) [reply packages/paste-website/src/components/shortcodes/table-of-contents/styles.ts]: As long as it is the exception - and isolated - I guess it's safe enough.  And I totally agree haha

File: packages/paste-website/src/pages/tokens/design-tokens-package.mdx (hunk 2 above)
Comment (richbachman) [reply packages/paste-website/src/pages/tokens/design-tokens-package.mdx]: Left over `id` from when I was adding the changelog anchor to the TOCs. Just forgot to remove.

File: packages/paste-website/src/components/shortcodes/table-of-contents/styles.ts (hunk 3 above)
Comment (richbachman) [reply packages/paste-website/src/components/shortcodes/table-of-contents/styles.ts]: I'm on board with that. I'll chat with @two24studios about looking at the design. Removing the style overrides for now.

File: packages/paste-website/src/components/shortcodes/table-of-contents/index.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import {TableOfContentsProps} from './types';
+import {StyledWrapper, StyledScrollSpy, StyledListItem} from './styles';
+import {slugify} from '../../../utils/RouteUtils';
+
+const TableOfContents: React.FC<TableOfContentsProps> = headings => {
+  if (headings == null || headings.data == null) {
+    return null;
+  }
+
+  const allHeadings = headings.data;
```
Comment (SiTaggart) [inline packages/paste-website/src/components/shortcodes/table-of-contents/index.tsx]: oh, so what I meant was change

```
<TableOfContents data={props.data.mdx.headings} />
```

to

```
<TableOfContents headings={props.data.mdx.headings} />
```

then you can destructure it from props

```
const TableOfContents: React.FC<TableOfContentsProps> = ({headings}) => {
  if (headings == null) {
    return null;
  }

  // Get Array of heading anchors. Excluding h1 elements.
  const headingsList = headings
    .filter(heading => heading.depth !== 1)
    .map(({value}) => {
      const headingAnchor = slugify(value);
      return headingAnchor;
    });
```

Comment (richbachman) [reply packages/paste-website/src/components/shortcodes/table-of-contents/index.tsx]: 👍 Got it.

**Pull request merged**